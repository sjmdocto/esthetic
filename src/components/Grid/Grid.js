import React, {useState} from 'react';
import {
  View,
  Text,
  FlatList,
  Image,
  Pressable,
  Modal,
  TouchableWithoutFeedback,
  TouchableOpacity,
} from 'react-native';
import filterCloset from '../../util/filterCloset';
import FilterContext from '../../util/FilterContext';
import {styles} from './Grid.style';

/**
 * Photo Grid for ClosetScreen
 * @function Grid
 * @param {*} props
 * @param {array} props.closet
 * @param {number} props.filterColor
 * @param {number} props.filterType
 * @param {function} props.onDelete
 */

const Grid = (props) => {
  const [detailVisible, setDetailVisible] = useState(false);
  const [selectionKey, setSelectionKey] = useState('');

  /**
   * @function onSelect
   * @param {number} key
   */
  const onSelect = (key) => {
    setSelectionKey(key);
    setDetailVisible(true);
  };
  /**
   * @function deleteHandler
   * @returns {void}
   */
  const deleteHandler = () => {
    props.onDelete(selectionKey);
    setDetailVisible(false);
  };

  /**
   * FlatList custom renderItem function
   * @function renderItem
   * @param {{object}} item
   */
  const renderItem = ({item}) => {
    return (
      <Pressable
        onLongPress={onSelect.bind(this, item.key)}
        accessible={true}
        accessibilityLabel={'Closet item'}>
        <View style={styles.item}>
          <Image
            source={{uri: `data:image/png;base64,${item.photo}`}}
            style={styles.photo}
          />
        </View>
      </Pressable>
    );
  };

  return (
    <FilterContext.Provider>
      <View style={styles.container}>
        <FlatList
          numColumns={3}
          data={filterCloset(props.closet, props.filterColor, props.filterType)}
          renderItem={renderItem}
          accessible={true}
        />
        <Modal visible={detailVisible} transparent={true}>
          <TouchableWithoutFeedback
            onPress={setDetailVisible.bind(this, false)}>
            <View style={styles.nonDetailView} />
          </TouchableWithoutFeedback>
          <View style={styles.detailView}>
            <TouchableOpacity
              onPress={setDetailVisible.bind(this, false)}
              style={styles.undoButton}>
              <View>
                <Text style={styles.buttonText}>Undo</Text>
              </View>
            </TouchableOpacity>
            <TouchableOpacity
              onPress={deleteHandler}
              style={styles.deleteButton}
              accessible={true}
              accessibilityLabel={'Delete button'}>
              <View>
                <Text style={styles.buttonText}>Delete</Text>
              </View>
            </TouchableOpacity>
          </View>
          <TouchableWithoutFeedback
            onPress={setDetailVisible.bind(this, false)}>
            <View style={styles.nonDetailView} />
          </TouchableWithoutFeedback>
        </Modal>
      </View>
    </FilterContext.Provider>
  );
};

export default Grid;
