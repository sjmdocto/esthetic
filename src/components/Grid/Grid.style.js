import common from '../../styles/common.style';
import {StyleSheet} from 'react-native';
import theme from '../../styles/theme.style';

export const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: theme.EST_GREY, // TODO: add this to theme
  },
  item: {
    margin: 1,
    flex: 1 / 3,
    height: common.SCREENWIDTH / 3,
    width: common.SCREENWIDTH / 3,
    // backgroundColor: 'black',
  },
  photo: {
    height: common.SCREENWIDTH / 3,
    width: common.SCREENWIDTH / 3,
  },
  nonDetailView: {
    flex: 3,
    // backgroundColor: 'red',
  },
  detailView: {
    flex: 1,
    flexDirection: 'row',
    alignItems: 'center',
    marginHorizontal: 20,
    // marginVertical: SCREENHEIGHT / 4,
    backgroundColor: theme.EST_LIGHT_BLACK,
    borderRadius: 30,
    //bottom: SCREENHEIGHT / 5,
    // borderWidth: 1,
    // borderColor: 'purple',
  },
  undoButton: {
    flex: 1,
    marginHorizontal: 20,
    // marginVertical: 5,
    height: 50,
    width: 1,
    backgroundColor: theme.EST_BLACK,
    borderRadius: 15,
    alignItems: 'center',
    justifyContent: 'center',
  },
  deleteButton: {
    flex: 1,
    marginHorizontal: 20,
    // marginVertical: 5,
    height: 50,
    width: 1,
    backgroundColor: 'darkred',
    borderRadius: 15,
    alignItems: 'center',
    justifyContent: 'center',
  },
  buttonText: {
    color: 'white',
    fontSize: 18,
  },
});
