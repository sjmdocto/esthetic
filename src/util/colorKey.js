export default {
  none: 'None',
  black: 'Black',
  white: 'White',
  grey: 'Grey',
  brown: 'Brown',
  red: 'Red',
  orange: 'Orange',
  yellow: 'Yellow',
  green: 'Green',
  blue: 'Blue',
  purple: 'Purple',
  pink: 'Pink',
  multicolored: 'Multicolored',
};
