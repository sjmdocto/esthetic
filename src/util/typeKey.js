export default {
  none: 0,
  outerwear: 1,
  formalwear: 2,
  sweaters: 3,
  hoodies: 4,
  buttonUps: 5,
  tees: 6,
  pants: 7,
  shorts: 8,
  activeWear: 9,
  shoes: 10,
};
