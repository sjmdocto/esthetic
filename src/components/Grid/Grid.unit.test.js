import React from 'react';
import Grid from './Grid';
import colorKey from '../../util/colorKey';
import typeKey from '../../util/typeKey';
import {render, fireEvent} from 'react-native-testing-library';

describe('Grid test suite:', () => {
  test('If closet is empty, no grid tiles are created.', () => {
    const emptyCloset = [];
    const {toJSON} = render(
      <Grid
        closet={emptyCloset}
        filterColor={colorKey.none}
        filterType={typeKey.none}
      />,
    );
    expect(toJSON()).toMatchSnapshot();
  });
  test('If closet is not empty, then grid tiles are created.', () => {
    const testCloset = [
      {
        photo: 'base64',
        colorTag: colorKey.black,
        typeTag: typeKey.outerwear,
        key: 1,
      },
      {
        photo: 'base64',
        colorTag: colorKey.black,
        typeTag: typeKey.outerwear,
        key: 2,
      },
      {
        photo: 'base64',
        colorTag: colorKey.black,
        typeTag: typeKey.pants,
        key: 3,
      },
      {
        photo: 'base64',
        colorTag: colorKey.black,
        typeTag: typeKey.pants,
        key: 4,
      },
    ];
    const {toJSON} = render(
      <Grid
        closet={testCloset}
        filterColor={colorKey.none}
        filterType={typeKey.none}
      />,
    );
    expect(toJSON()).toMatchSnapshot();
  });
  test('Delete button is functional.', () => {
    const testCloset = [
      {
        photo: 'base64',
        colorTag: colorKey.black,
        typeTag: typeKey.outerwear,
        key: 1,
      },
      {
        photo: 'base64',
        colorTag: colorKey.black,
        typeTag: typeKey.outerwear,
        key: 2,
      },
      {
        photo: 'base64',
        colorTag: colorKey.black,
        typeTag: typeKey.pants,
        key: 3,
      },
      {
        photo: 'base64',
        colorTag: colorKey.black,
        typeTag: typeKey.pants,
        key: 4,
      },
    ];
    const mockOnDelete = jest.fn().mockName('onDelete');
    const {getByLabelText} = render(
      <Grid
        closet={testCloset}
        filterColor={colorKey.none}
        filterType={typeKey.none}
        onDelete={mockOnDelete}
      />,
    );
    fireEvent.press(getByLabelText('Delete button'));
    expect(mockOnDelete).toHaveBeenCalled();
  });
});
