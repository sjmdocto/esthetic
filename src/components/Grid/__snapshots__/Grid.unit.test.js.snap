// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Grid test suite: If closet is empty, no grid tiles are created. 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#E5E5E5",
      "flex": 1,
    }
  }
>
  <RCTScrollView
    accessible={true}
    data={Array []}
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <View />
  </RCTScrollView>
  <Modal
    hardwareAccelerated={false}
    transparent={true}
    visible={false}
  >
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "flex": 3,
        }
      }
    />
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#1F1F1F",
          "borderRadius": 30,
          "flex": 1,
          "flexDirection": "row",
          "marginHorizontal": 20,
        }
      }
    >
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#141414",
            "borderRadius": 15,
            "flex": 1,
            "height": 50,
            "justifyContent": "center",
            "marginHorizontal": 20,
            "opacity": 1,
            "width": 1,
          }
        }
      >
        <View>
          <Text
            style={
              Object {
                "color": "white",
                "fontSize": 18,
              }
            }
          >
            Undo
          </Text>
        </View>
      </View>
      <View
        accessibilityLabel="Delete button"
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "darkred",
            "borderRadius": 15,
            "flex": 1,
            "height": 50,
            "justifyContent": "center",
            "marginHorizontal": 20,
            "opacity": 1,
            "width": 1,
          }
        }
      >
        <View>
          <Text
            style={
              Object {
                "color": "white",
                "fontSize": 18,
              }
            }
          >
            Delete
          </Text>
        </View>
      </View>
    </View>
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "flex": 3,
        }
      }
    />
  </Modal>
</View>
`;

exports[`Grid test suite: If closet is not empty, then grid tiles are created. 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#E5E5E5",
      "flex": 1,
    }
  }
>
  <RCTScrollView
    accessible={true}
    data={
      Array [
        Object {
          "colorTag": 1,
          "key": 1,
          "photo": "base64",
          "typeTag": 1,
        },
        Object {
          "colorTag": 1,
          "key": 2,
          "photo": "base64",
          "typeTag": 1,
        },
        Object {
          "colorTag": 1,
          "key": 3,
          "photo": "base64",
          "typeTag": 2,
        },
        Object {
          "colorTag": 1,
          "key": 4,
          "photo": "base64",
          "typeTag": 2,
        },
      ]
    }
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            Object {
              "flexDirection": "row",
            }
          }
        >
          <View
            accessibilityLabel="Closet item"
            accessible={true}
            focusable={true}
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
          >
            <View
              style={
                Object {
                  "flex": 0.3333333333333333,
                  "height": 250,
                  "margin": 1,
                  "width": 250,
                }
              }
            >
              <Image
                source={
                  Object {
                    "uri": "data:image/png;base64,base64",
                  }
                }
                style={
                  Object {
                    "height": 250,
                    "width": 250,
                  }
                }
              />
            </View>
          </View>
          <View
            accessibilityLabel="Closet item"
            accessible={true}
            focusable={true}
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
          >
            <View
              style={
                Object {
                  "flex": 0.3333333333333333,
                  "height": 250,
                  "margin": 1,
                  "width": 250,
                }
              }
            >
              <Image
                source={
                  Object {
                    "uri": "data:image/png;base64,base64",
                  }
                }
                style={
                  Object {
                    "height": 250,
                    "width": 250,
                  }
                }
              />
            </View>
          </View>
          <View
            accessibilityLabel="Closet item"
            accessible={true}
            focusable={true}
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
          >
            <View
              style={
                Object {
                  "flex": 0.3333333333333333,
                  "height": 250,
                  "margin": 1,
                  "width": 250,
                }
              }
            >
              <Image
                source={
                  Object {
                    "uri": "data:image/png;base64,base64",
                  }
                }
                style={
                  Object {
                    "height": 250,
                    "width": 250,
                  }
                }
              />
            </View>
          </View>
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            Object {
              "flexDirection": "row",
            }
          }
        >
          <View
            accessibilityLabel="Closet item"
            accessible={true}
            focusable={true}
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
          >
            <View
              style={
                Object {
                  "flex": 0.3333333333333333,
                  "height": 250,
                  "margin": 1,
                  "width": 250,
                }
              }
            >
              <Image
                source={
                  Object {
                    "uri": "data:image/png;base64,base64",
                  }
                }
                style={
                  Object {
                    "height": 250,
                    "width": 250,
                  }
                }
              />
            </View>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
  <Modal
    hardwareAccelerated={false}
    transparent={true}
    visible={false}
  >
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "flex": 3,
        }
      }
    />
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#1F1F1F",
          "borderRadius": 30,
          "flex": 1,
          "flexDirection": "row",
          "marginHorizontal": 20,
        }
      }
    >
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#141414",
            "borderRadius": 15,
            "flex": 1,
            "height": 50,
            "justifyContent": "center",
            "marginHorizontal": 20,
            "opacity": 1,
            "width": 1,
          }
        }
      >
        <View>
          <Text
            style={
              Object {
                "color": "white",
                "fontSize": 18,
              }
            }
          >
            Undo
          </Text>
        </View>
      </View>
      <View
        accessibilityLabel="Delete button"
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "darkred",
            "borderRadius": 15,
            "flex": 1,
            "height": 50,
            "justifyContent": "center",
            "marginHorizontal": 20,
            "opacity": 1,
            "width": 1,
          }
        }
      >
        <View>
          <Text
            style={
              Object {
                "color": "white",
                "fontSize": 18,
              }
            }
          >
            Delete
          </Text>
        </View>
      </View>
    </View>
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "flex": 3,
        }
      }
    />
  </Modal>
</View>
`;
